src/database/rusty.rs

```diff
@@ -93,54 +98,174 @@ where
     }
 }
 
-impl<Key: Serialize + DeserializeOwned, Value: Serialize + DeserializeOwned>
-    RustyLevelDB<Key, Value>
-{
-    pub fn new_iter(&self) -> RustyLevelDBIterator<Key, Value> {
-        RustyLevelDBIterator::new(self)
-    }
-
-    pub fn flush(&mut self) {
-        self.database
-            .write(&WriteBatch::new(), true)
-            .expect("Database flushing to disk must succeed");
-    }
-}
+/// `RustyLevelDbAsync` provides an async-friendly and clone-friendly wrapper
+/// around [`RustyLevelDB`].
+///
+/// Methods in the underlying struct `LevelDB` from `rs-leveldb` crate are all sync
+/// and they sometimes perfom blocking file IO.  It is discouraged to
+/// call blocking IO from async code as it can lead to concurrency problems,
+/// usually hidden until a certain level of load is reached.
+///
+/// The tokio page for spawn_blocking says:
+///     In general, issuing a blocking call or performing a lot of compute in a
+///     future without yielding is problematic, as it may prevent the executor from
+///     driving other futures forward.
+///
+/// See:
+///  * <https://github.com/Neptune-Crypto/neptune-core/issues/74>
+///  * <https://internals.rust-lang.org/t/warning-when-calling-a-blocking-function-in-an-async-context/11440/5>
+///  * <https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html>
+///
+/// Therefore `RustyLevelDbAsync` wraps the sync methods with `spawn_blocking()`
+/// so that the tokio runtime can run the blocking IO on a thread where blocking
+/// is acceptable
```

Comment:
I'm a bit unsure about why this is needed. The general struct seems to be 

```rust
pub async fn <method>(&self, other_args) {
    let inner = self.0.clone();
    task::spawn_blocking(move || inner.<method>(other_args)).await.unwrap()
}
```

So you're just running the database operation in a new thread and waiting for that thread to finish instead of just running it in the original thread? I don't fully understand why that is beneficial. Maybe you can elaborate on what the documentation from `tokio` says as I'm not sure I fully get it?

Reply:
I will try.   You probably know most of this, but anyway...

So the purpose of async/await functions (futures) is that they can return immediately to the caller without blocking.  The executor then polls the function to see if it is ready yet or not.   By doing this with multiple functions, it can effectively "multitask" performing multiple operations concurrently, even from a single OS thread.  Unlike native threads that can be pre-empted by the OS, this is a cooperative execution model.   It requires voluntary participation.    So an async function that calls blocking code is basically not cooperating.  It is blocking the entire *native/OS* thread, which prevents the executor from polling (aka driving forward) other tasks.  So then, concurrency suffers.    This is why crates like tokio provide async replacements for many std lib items, especially the filesystem APIs which are written in a blocking manner.

There is some further nuance here.  The default tokio runtime (which we are using) uses a multi-threaded executor.   By default (again) it creates one thread per CPU, eg on my laptop that's 4 native OS threads.   So even if one is blocked, some tasks can still proceed, which may alleviate the problem to a degree.   I do not know the details beyond that.... it may be possible for a single blocking function to still stall some other task(s).  But let's say we have 4 async functions executing simultaneously, one on each CPU and each is calling a blocking function.  Then all other tasks are blocked.   oops.

So, what it boils down to is:  async best-practice is to avoid calling blocking functions, especially file I/O.  File I/O takes a really long time and most of it does not involve the CPU, so that's just wasted capacity.

So that's why I filed #74 and #75 a while back.

Now, with regards to the spawn_blocking() calls in `RustyLevelDBAsync` that is basically a way to inform the tokio executor "hey I'm going to do some blocking stuff, please prepare accordingly".  So tokio may create a new OS thread, or move my task to an existing thread in a thread pool.  Anyway, it arranges to keep its regular OS threads freed up for polling/advancing well-behaving futures that return quickly.

I may be off on a detail here or there, but I believe the overall gist is correct.